src/common/hist_builder.h

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright 2017-2021 by Contributors
+ * \file hist_builder.h
+ */
+#ifndef XGBOOST_COMMON_HIST_BUILDER_H_
+#define XGBOOST_COMMON_HIST_BUILDER_H_
+
+#include <algorithm>
+#include <vector>
+#include "hist_util.h"
+#include "../data/gradient_index.h"
+
+#if defined(XGBOOST_MM_PREFETCH_PRESENT)
+  #include <xmmintrin.h>
+  #define PREFETCH_READ_T0(addr) _mm_prefetch(reinterpret_cast<const char*>(addr), _MM_HINT_T0)
+#elif defined(XGBOOST_BUILTIN_PREFETCH_PRESENT)
+  #define PREFETCH_READ_T0(addr) __builtin_prefetch(reinterpret_cast<const char*>(addr), 0, 3)
+#else  // no SW pre-fetching available; PREFETCH_READ_T0 is no-op
+  #define PREFETCH_READ_T0(addr) do {} while (0)
+#endif  // defined(XGBOOST_MM_PREFETCH_PRESENT)
+
+namespace xgboost {
+namespace common {
+
+struct Prefetch {
+ public:
+  static constexpr size_t kCacheLineSize = 64;
+  static constexpr size_t kPrefetchOffset = 10;
+
+ private:
+  static constexpr size_t kNoPrefetchSize =
+      kPrefetchOffset + kCacheLineSize /
+      sizeof(decltype(GHistIndexMatrix::row_ptr)::value_type);
+
+ public:
+  static size_t NoPrefetchSize(size_t rows) {
+    return std::min(rows, kNoPrefetchSize);
+  }
+
+  template <typename T>
+  static constexpr size_t GetPrefetchStep() {
+    return Prefetch::kCacheLineSize / sizeof(T);
+  }
+};
+
+template<typename FPType, bool do_prefetch,
+         typename BinIdxType, bool is_root,
+         bool any_missing>
+void BuildHistKernel(const std::vector<GradientPair>& gpair,
+                     const uint32_t* rows,
+                     const uint32_t row_begin,
+                     const uint32_t row_end,
+                     const GHistIndexMatrix& gmat,
+                     const BinIdxType* numa_data,
+                     uint16_t* nodes_ids,
```

Comment:
Could you please share why do you need to pass nodes_id for building histogram instead of passing the histogram for this particular node?  This way we can have a small scope for this function and makes it easier to test.

Also, I think the node should have 32-bit type `bst_node_t`.

Reply:
Sure.
Changes in `hist_builder.h` were introduced due to changes in partition builder.
Before optimizations "full" bin matrix went through the memory for each node on the same level during histogram building: potentially for each node on specific level we have to read the bin matrix almost from the beginning to the end with random strides. The bandwidth of the memory is significantly dropped in such conditions.
So to mitigate this problem in optimized code we have "sorted" row indecies (but no sorting is done exactly! :) ) for each specific level and as a side effect (rows are not grouped by nodes) we have to know what `node_id` specific row is belong to. `node_id` is required to know what node's histogram should be updated with specific bin matrix row.

Sorry if description is not clear, please let me know and I'll try to add more details.

As for 32-bit type usage, now it works for max_depth <= 16 perfectly. I have to check the performance impact before changing it from 16 to 32 bit and I hope there won't be any reason to add template parameter for it.

Reply:
Thank you for the detailed explanation!  That is an insightful optimization.

For future references with these types of optimization, I think you should put that on the front page of PR before anything else and in the code comment, and potentially open an RFC like https://github.com/dmlc/xgboost/issues/7308 before making the change, so we can be clear about the design choice. (I write design doc from time to time just to keep myself clear about the changes).  For this PR, I asked myself again, there's no way I can guess what you are trying to do by looking at the code.

So, let me try to summarize your summary ;-)  Feel free to correct me if I'm wrong.  Instead of asking which row should we process given a node index, you ask which node should we process given a row index.  This optimization is only applicable when the depth-wise grow policy is used, so the code is split and specialized.

Can you design an intuitive interface for the partitioner and the histogram builder so that the code can be easily reused?  I can give it some thoughts if you can confirm my summary.

> As for 32-bit type usage, now it works for max_depth <= 16 perfectly.

I don't think this optimization is necessary.

Reply:
> Instead of asking which row should we process given a node index, you ask which node should we process given a row index. This optimization is only applicable when the depth-wise grow policy is used, so the code is split and specialized.

It's clear and right summary :)
Yes. I had to specialize the partition builder related code for depth-wise and loss-guided grow polices, as for loss-guided policy we have to know exactly the set of rows splited to left and to right child nodes.

>Can you design an intuitive interface for the partitioner and the histogram builder so that the code can be easily reused? I can give it some thoughts if you can confirm my summary.

Would it be applicable if the `OptPartitionBuilder` class supported the `depthwise` and `lossguided` policies (strategy pattern) instead of `is_loss_guided` template usage?
As for histogram builder, there is no significant changes in `BuildHistKernel`, only `node_id` usage which is applicable for both growing policies.